src/middleware/cors/service.rs

```diff
@@ -1,41 +1,211 @@
-use super::Config;
+use super::{Config, CorsResource};
 
-use futures::{Poll};
-use http;
+use futures::{Async, Future, Poll};
+use http::{self, HeaderMap, StatusCode};
+use http::{Request, Response};
 use tower_service::Service;
 
 use std::sync::Arc;
 
+#[derive(Debug)]
 pub struct CorsService<S> {
     inner: S,
     config: Arc<Config>,
 }
 
 impl<S> CorsService<S> {
     pub(super) fn new(inner: S, config: Arc<Config>) -> CorsService<S> {
-        CorsService {
-            inner,
-            config,
-        }
+        CorsService { inner, config }
     }
 }
 
-use util::http::HttpService;
-use http::{Request, Response};
-
 impl<S> Service for CorsService<S>
-where S: HttpService,
+where
+    S: ::util::http::HttpService,
 {
     type Request = Request<S::RequestBody>;
-    type Response = Response<S::ResponseBody>;
+    type Response = Response<Option<S::ResponseBody>>;
     type Error = S::Error;
-    type Future = S::Future;
+    type Future = CorsFuture<S::Future>;
 
     fn poll_ready(&mut self) -> Poll<(), Self::Error> {
         self.inner.poll_ready()
     }
 
     fn call(&mut self, request: Self::Request) -> Self::Future {
-        self.inner.call(request)
+        match self.config.process_request(&request) {
+            Ok(CorsResource::Preflight(headers)) => CorsFuture::Handled(Some(headers)),
+            Ok(CorsResource::Simple(headers)) => {
+                CorsFuture::Simple(self.inner.call(request), Some(headers))
+            }
+            Err(e) => {
+                eprintln!("{:?}", e);
+                CorsFuture::Handled(None)
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum CorsFuture<F> {
+    Simple(F, Option<HeaderMap>),
+    Handled(Option<HeaderMap>),
+}
+
+impl<F, ResponseBody> Future for CorsFuture<F>
+where
+    F: Future<Item = http::Response<ResponseBody>>,
+{
+    type Item = http::Response<Option<ResponseBody>>;
+    type Error = F::Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        use self::CorsFuture::*;
+
+        match self {
+            Simple(f, headers) => {
+                let mut response = try_ready!(f.poll());
+                let headers = headers.take().expect("poll called twice");
+                response.headers_mut().extend(headers);
+                Ok(Async::Ready(response.map(Some)))
+            }
+            Handled(headers) => {
+                let mut response = http::Response::new(None);
+                *response.status_mut() = StatusCode::BAD_REQUEST;
```

Comment:
If we're going to return a response directly for a CORS violation, the [spec](https://fetch.spec.whatwg.org/#http-cors-protocol) suggests sending a 403 status code.